File: prototype/rails/src/app/domain/shop/favorite_change.rb
require_relative "../listings/listing_event_type"

module Domain
  module Shop
    # One button favorites and unfavorites, so what it does next follows from
    # what the visitor has saved already.
    module FavoriteChange
      ADDED = "added"
      REMOVED = "removed"

      ALL = [ADDED, REMOVED].freeze

      LISTING_EVENTS = {
        ADDED => Listings::ListingEventType::FAVORITE,
        REMOVED => Listings::ListingEventType::UNFAVORITE
      }.freeze

      module_function

      def from_current_state(favorited)
        favorited ? REMOVED : ADDED
      end

      def listing_event(change)
        LISTING_EVENTS.fetch(change)
      end

      def added?(change)
        change == ADDED
      end
    end
  end
end

File: prototype/node/src/app/core/listings/favorite-change.ts
import type { ListingEventType } from './listing-event-type.ts'

export const FAVORITE_CHANGES = ['added', 'removed'] as const
export type FavoriteChange = (typeof FAVORITE_CHANGES)[number]

const LISTING_EVENTS: Record<FavoriteChange, ListingEventType> = {
  added: 'favorite',
  removed: 'unfavorite',
}

// One button both favorites and unfavorites, so what it does next follows
// from what the visitor has saved already.
export function favoriteChangeFor(isFavorited: boolean): FavoriteChange {
  return isFavorited ? 'removed' : 'added'
}

export function listingEventForFavoriteChange(change: FavoriteChange): ListingEventType {
  return LISTING_EVENTS[change]
}
